utils/stall_free.hh

```diff
@@ -57,14 +59,176 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+// The clear_gently functions are meant for
+// gently destroying the contents of containers.
+// The containers can be re-used after clear_gently
+// or may be destroyed.  But unlike e.g. std::vector::clear(),
+// clear_gently will not necessarily keep the object allocation.
+
+template <typename T>
+concept HasClearGentlyMethod = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept SharedPointer = SmartPointer<T> && requires (T x) {
+    { x.use_count() } -> std::convertible_to<long>;
+};
+
+template <typename T>
+concept StringLike = requires (T x) {
+    std::is_same_v<typename T::traits_type, std::char_traits<typename T::value_type>>;
+};
+
+template <typename T>
+concept Iterable = requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.begin() } -> std::same_as<typename T::iterator>;
+    { x.end() } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept Sequence = Iterable<T> && requires (T x, size_t n) {
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept TriviallyClearableSequence =
+    Sequence<T> && std::is_trivially_destructible_v<typename T::value_type> && !HasClearGentlyMethod<typename T::value_type>;
```

Comment:
Doesn't a "trivially clearable" sequence require all that **and** also a clear() method? If not, now is it clearable?

Reply:
This was meant for it to be Clearable by us, i.e. can have a trivial `gently_clear` implementation.
I intentionally chose not to call `clear()` in this case since it will not free the instance's allocation, while in the scylla's use cases we have at hand (token_metadata and repair) we do want to deallocate the contents of containers, including vectors and std::vector::clear just doesn't do that.
I don't want to be constrained by semantics. Maybe `clear_gently` isn't a perfect choice, but we already use it for this purpose.